File: examples/quickstart/python/pulserpc/transport.py
"""Transport abstraction for PulseRPC clients"""

from abc import ABC, abstractmethod
from typing import Any, Dict
import json
import urllib.request
import urllib.error


class Transport(ABC):
    """Abstract base class for RPC transports"""

    @abstractmethod
    def request(self, req: Dict[str, Any]) -> Dict[str, Any]:
        """Send a JSON-RPC request and return the response

        Args:
            req: JSON-RPC request dict with 'jsonrpc', 'method', 'params', 'id'

        Returns:
            JSON-RPC response dict

        Raises:
            urllib.error.URLError: For HTTP transport errors
            ValueError: For invalid responses
        """
        pass


class HttpTransport(Transport):
    """HTTP transport implementation using urllib"""

    def __init__(self, url: str, timeout: int = 30, headers: Dict[str, str] = None):
        """Initialize HTTP transport

        Args:
            url: RPC endpoint URL
            timeout: Request timeout in seconds (default: 30)
            headers: Optional HTTP headers to include with requests
        """
        self.url = url
        self.timeout = timeout
        self.headers = headers or {}

    def request(self, req: Dict[str, Any]) -> Dict[str, Any]:
        """Send request via HTTP POST

        Args:
            req: JSON-RPC request dict

        Returns:
            JSON-RPC response dict

        Raises:
            urllib.error.URLError: For network/HTTP errors
            ValueError: For invalid JSON responses
        """
        # Prepare request body
        body = json.dumps(req).encode('utf-8')

        # Create request
        http_req = urllib.request.Request(
            self.url,
            data=body,
            method='POST',
            headers={
                'Content-Type': 'application/json',
                **self.headers
            }
        )

        # Send request and get response
        with urllib.request.urlopen(http_req, timeout=self.timeout) as response:
            response_body = response.read().decode('utf-8')

            # Handle empty response (notification)
            if not response_body:
                return {}

            # Parse JSON response
            try:
                return json.loads(response_body)
            except json.JSONDecodeError as e:
                raise ValueError(f"Invalid JSON response: {e}") from e


class InProcTransport(Transport):
    """In-process transport for testing (directly calls Server)"""

    def __init__(self, server):
        """Initialize in-process transport

        Args:
            server: Server instance to call directly
        """
        self.server = server

    def request(self, req: Dict[str, Any]) -> Dict[str, Any]:
        """Send request directly to Server instance

        Args:
            req: JSON-RPC request dict

        Returns:
            JSON-RPC response dict (or empty dict for notifications)
        """
        ctx = req.get('ctx')
        response = self.server.call(req, ctx)
        return response if response is not None else {}

File: pkg/runtime/runtimes/ts-cjs/pulserpc/transport.ts
/**
 * Transport abstraction for JSON-RPC requests.
 *
 * Provides abstract Transport base class and concrete implementations:
 * - HttpTransport: HTTP-based transport using fetch
 * - InProcTransport: In-process transport for testing
 */

import type { JsonRpcRequest, JsonRpcResponse } from "./types";
import { RPCError } from "./rpc";

/**
 * Abstract transport base class
 */
export class Transport {
  async request(_req: JsonRpcRequest): Promise<JsonRpcResponse> {
    throw new Error("Transport.request must be implemented by subclass");
  }

  close(): void {
    // Default implementation does nothing
  }
}

/**
 * HTTP transport implementation using fetch
 */
export class HttpTransport extends Transport {
  baseUrl: string;
  headers: Record<string, string>;
  fetchFn: typeof fetch;

  constructor(baseUrl: string, headers: Record<string, string> = {}, fetchFn: typeof fetch = fetch) {
    super();
    // The endpoint URL is used verbatim; do not strip or append anything so
    // callers can target a specific path (e.g. a reverse-proxied /api/rpc).
    this.baseUrl = baseUrl;
    this.headers = {
      "Content-Type": "application/json",
      ...headers,
    };
    this.fetchFn = fetchFn;
  }

  async request(req: JsonRpcRequest): Promise<JsonRpcResponse> {
    const url = this.baseUrl;

    const response = await this.fetchFn(url, {
      method: "POST",
      headers: this.headers,
      body: JSON.stringify(req),
    });

    if (!response.ok) {
      throw new RPCError(-32000, `HTTP ${response.status}: ${response.statusText}`, { status: response.status });
    }

    const text = await response.text();
    if (!text) {
      throw new RPCError(-32000, "Empty response body");
    }

    try {
      return JSON.parse(text);
    } catch (e: any) {
      throw new RPCError(-32000, `Invalid JSON response: ${e.message}`);
    }
  }
}

/**
 * In-process transport for testing.
 *
 * This transport calls the server directly without HTTP.
 */
export class InProcTransport extends Transport {
  serverCallFn: (req: JsonRpcRequest) => JsonRpcResponse | undefined;

  constructor(serverCallFn: (req: JsonRpcRequest) => JsonRpcResponse | undefined) {
    super();
    this.serverCallFn = serverCallFn;
  }

  async request(req: JsonRpcRequest): Promise<JsonRpcResponse> {
    const response = this.serverCallFn(req);
    if (!response) {
      return {
        jsonrpc: "2.0",
        id: req.id ?? null,
      };
    }
    return response;
  }
}
